programs/bpf_loader/src/lib.rs

```diff
@@ -808,85 +856,78 @@ fn process_loader_upgradeable_instruction(
+            {
                 ic_logger_msg!(
                     log_collector,
                     "Recipient is the same as the account being closed"
                 );
                 return Err(InstructionError::InvalidArgument);
             }
+            let mut close_account = instruction_context
+                .try_borrow_account(transaction_context, first_instruction_account)?;
 
-            match close_account.state()? {
+            match close_account.get_state()? {
                 UpgradeableLoaderState::Uninitialized => {
-                    recipient_account
-                        .try_account_ref_mut()?
-                        .checked_add_lamports(close_account.lamports()?)?;
-                    close_account.try_account_ref_mut()?.set_lamports(0);
+                    let mut recipient_account = instruction_context
+                        .try_borrow_account(transaction_context, first_instruction_account + 1)?;
+                    recipient_account.checked_add_lamports(close_account.get_lamports())?;
+                    close_account.set_lamports(0);
 
                     ic_logger_msg!(
                         log_collector,
                         "Closed Uninitialized {}",
-                        close_account.unsigned_key()
+                        close_account.get_key()
                     );
                 }
                 UpgradeableLoaderState::Buffer { authority_address } => {
-                    let authority =
-                        keyed_account_at_index(keyed_accounts, first_instruction_account + 2)?;
-
+                    drop(close_account);
                     common_close_account(
                         &authority_address,
-                        authority,
-                        close_account,
-                        recipient_account,
+                        transaction_context,
+                        instruction_context,
+                        first_instruction_account,
                         &log_collector,
                     )?;
 
-                    ic_logger_msg!(
-                        log_collector,
-                        "Closed Buffer {}",
-                        close_account.unsigned_key()
-                    );
+                    let close_account = instruction_context
+                        .try_borrow_account(transaction_context, first_instruction_account)?;
+                    ic_logger_msg!(log_collector, "Closed Buffer {}", close_account.get_key());
                 }
                 UpgradeableLoaderState::ProgramData {
                     slot: _,
                     upgrade_authority_address: authority_address,
                 } => {
-                    let program_account =
-                        keyed_account_at_index(keyed_accounts, first_instruction_account + 3)?;
-
-                    if !program_account.is_writable() {
+                    if !instruction_context.is_writable(first_instruction_account + 3)? {
```

Comment:
starting to have too many magic numbers here, would be nice to have:
`let program_index = first_instruction_account + 3;`